ftprime/argrecorder.py

```diff
@@ -51,197 +120,92 @@ def add_record(self, left, right, parent, children):
         [left,right).
         '''
         # unneeded but helpful for debugging
-        if parent not in self.keys():
+        out_parent = self.node_ids[parent - self.input_min]
+        if out_parent == NULL_ID:
             raise ValueError("Parent " + str(parent) +
                              "'s birth time has not been recorded with " +
                              ".add_individual().")
-        # time = self[parent][0]
-        new_rec = msprime.Edgeset(
-                parent=parent,
-                children=children,
-                left=left,
-                right=right)
-        merge_records(new_rec, self[parent][1])
-
-    def edgesets(self):
-        '''
-        Returns an iterator of Edgesets, sorted in reverse order by
-        the order the parents were input; so if parents are input in order
-        by birth time, the output will be in the time order required by
-        msprime.
-        '''
-        for y in reversed(self.values()):
-            for z in y[1]:
-                yield z
+        out_children = tuple([self.node_ids[u - self.input_min] for u in children])
+        self.edgesets.add_row(parent=out_parent,
+                              children=out_children,
+                              left=left,
+                              right=right)
 
-    def edgeset_table(self):
-        '''
-        Returns an EdgesetTable instance.
-        '''
-        table = msprime.EdgesetTable()
-        for es in self.edgesets():
-            table.add_row(
-                left=es.left, right=es.right, parent=es.parent,
-                children=es.children)
-        return table
-
-    def node_table(self):
-        '''
-        Returns a NodeTable instance, whose k-th row corresponds to node k.
-        Translates time to 'time ago' by subtracting time from max_time, which
-        by default is the largest time seen so far.
-        '''
-        table = msprime.NodeTable()
-        empty_node = msprime.Node(time=0.0)
-        for k in range(self.num_nodes):
-            try:
-                node = self[k][0]
-            except KeyError:
-                node = empty_node
-            table.add_row(flags=node.flags, time=self.max_time - node.time,
-                          population=node.population)
-        return table
+    def add_mutation(self, position, node, derived_state):
+        """
+        Adds a new mutation to mutation table, and a new site if necessary as well.
+        """
+        if position not in self.sites.position:
```

Comment:
oh!  damn.  I was imagining the numpy stuff as accessing the internals of the data structure.  heh.

I've added a note to the documentation to explain this point.

I've just been tacking everything onto the end of tables and letting `sort_tables` sort it out: is this reasonable, or would it be better to maintain a heapq or something?